ambry-frontend/src/main/java/com/github/ambry/frontend/AmbryIdConverterFactory.java

```diff
@@ -112,37 +113,97 @@ public Future<String> convert(RestRequest restRequest, String input, BlobInfo bl
           convertedId = "/" + signIdIfRequired(restRequest, input);
         } else {
           CallbackUtils.callCallbackAfter(convertId(input, restRequest, blobInfo),
-              (id, e) -> completeConversion(id, e, future, callback));
+              (id, e) -> completeConverterOperation(id, e, future, callback,
+                  frontendMetrics.idConversionDownstreamCallbackTimeInMs));
         }
       } catch (Exception e) {
         exception = e;
       } finally {
         frontendMetrics.idConverterProcessingTimeInMs.update(System.currentTimeMillis() - startTimeInMs);
         if (convertedId != null || exception != null) {
-          completeConversion(convertedId, exception, future, callback);
+          completeConverterOperation(convertedId, exception, future, callback, frontendMetrics.idConversionDownstreamCallbackTimeInMs);
         }
       }
       return future;
     }
 
+    @Override
+    public Future<Boolean> detectConflict(RestRequest restRequest, Callback<Boolean> callback) {
+      final CompletableFuture<Boolean> future = new CompletableFuture<>();
+      Exception exception = null;
+      frontendMetrics.idConverterDetectConflictRate.mark();
+      long startTimeInMs = System.currentTimeMillis();
+      try {
+        if (!isOpen) {
+          exception = new RestServiceException("IdConverter is closed", RestServiceErrorCode.ServiceUnavailable);
+        } else {
+          CallbackUtils.callCallbackAfter(detectConflict(restRequest),
+              (conflictStatus, e) -> completeConverterOperation(conflictStatus, e, future, callback, frontendMetrics.idConversionDetectConflictDownstreamCallbackTimeInMs));
+        }
+      } catch (Exception e) {
+        exception = e;
+      } finally {
+        frontendMetrics.deleteCallbackProcessingTimeInMs.update(System.currentTimeMillis() - startTimeInMs);
+        if (exception != null) {
+          completeConverterOperation(false, exception, future, callback, frontendMetrics.idConversionDetectConflictDownstreamCallbackTimeInMs);
+        }
+      }
+      return future;
+    }
+
+    /**
+     * Detects if the ID in the request can cause a conflict.
+     * @param restRequest @link RestRequest} representing the request.
+     * @return the {@link CompletionStage} that will be completed with the conflict status.
+     * @throws RestServiceException if the operation cannot be performed.
+     */
+    private CompletionStage<Boolean> detectConflict(RestRequest restRequest) throws RestServiceException {
+      final CompletionStage<Boolean> detectionFuture;
+      frontendMetrics.idConverterDetectConflictRate.mark();
+      if (restRequest.getRestMethod() == RestMethod.PUT && RestUtils.getRequestPath(restRequest)
```

Comment:
Can we verify the Dataset behavior?
Does it allow update with the same version number if upsert is not set?